[ACS] BREAKING CHANGE: ARO remove vnet-peer, upgrade to 2019-10-27-preview API version

olga-mir opened the pull request:
This is a draft PR to allow early feedback. 
This PR updates ARO module to latest API version `2019-10-27-preview`.  It depends on https://github.com/Azure/azure-rest-api-specs/pull/8305 (in the process of being merged) and the consequent sdk-for-python build. 
In the new version vnet-peer has been removed from "create" operation. This is due to latest changes done on Azure side while implementing new API features it is now possible to edit vnet peering after the cluster creation. I have tried deprecation approach (https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#deprecating-commands-and-arguments) but unfortunately in this circumstances we have to remove it now rather than adding a deprecation notice. 

**History Notes:**  
[ACS] BREAKING CHANGE: az openshift create: remove --vnet-peer parameter.  
[ACS] az openshift create: add flags to support private cluster. 
[ACS] az openshift: upgrade to `2019-10-27-preview` API version.
[ACS] az openshift: add `update` command.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

fengzhou-msft commented:
Updated history notes section to conform to the [required format](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

fengzhou-msft commented on src/azure-cli/azure/cli/command_modules/acs/_help.py:
```diff
@@ -1002,3 +969,16 @@
+parameters:
+  - name: --refresh-cluster
+    type: boolean
+    short-summary: Trigger nodes rotation in order to pick up change in DNS settings.
```
You already have the help text in _params.py. You can remove it here.

fengzhou-msft commented on src/azure-cli/azure/cli/command_modules/acs/_params.py:
```diff
@@ -324,9 +324,25 @@ def load_arguments(self, _):
+        c.argument('workspace_id', help='The resource id of an existing Log Analytics Workspace to use for storing monitoring data.')
+
+    with self.argument_context('openshift update') as c:
+        c.argument('refresh_cluster', arg_type=get_three_state_flag(),
+                   help='Allow nodes to be rotated. Use this flag to trigger nodes rotation after DNS settings change.')
```
I suggest to use `action='store_true'` instead of three_state_flag since `refresh_cluster` is not a property of the resource. See more context [here](https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#registering-flags).

olga-mir replied:
@fengzhou-msft thank you for reviewing and improvement suggestions. I have implemented them locally but lately there were few problems with RP that prevented from testing my changes. I will hold off for now working on this PR as we are still waiting for the core python sdk to be updated and then I'll test everything together.

olga-mir replied:
@fengzhou-msft we do store this flag on the resource properties: https://github.com/openshift/openshift-azure/blob/3f03631cc8a435afde90b158289613ee4140b07e/pkg/api/admin/types.go#L75
It is not returned in the response payload or in subsequent `az openshift show` commands. should I still use `store_true` rather than `three_state_flag`?

fengzhou-msft replied:
If it's a resource property, then `get_three_state_flag` is the right choice.

kwoodson commented:
@olga-mir Could you please update this with your status?  Thanks!

olga-mir commented:
hi @kwoodson this can't be merged yet because the core sdk has not been updated with the latest API specs. On the last scrum with MSFT last Friday (6th March) I've been told that there is a conflict with AKS core sdk release/versioning and it's been taking a lot of time to resolve. I will resume working on this task when this has been sorted out

ehashman commented on src/azure-cli/azure/cli/command_modules/acs/custom.py:
```diff
@@ -66,15 +66,17 @@
+from azure.mgmt.containerservice.v2019_10_27_preview.models import OpenShiftManagedClusterAuthProfile
+from azure.mgmt.containerservice.v2019_10_27_preview.models import NetworkProfile
+from azure.mgmt.containerservice.v2019_10_27_preview.models import OpenShiftManagedClusterMonitorProfile
+from azure.mgmt.containerservice.v2019_10_27_preview.models import OpenShiftAPIProperties
```
Per https://github.com/Azure/sdk-release-request/issues/46 this should be available in the latest release (8.3.0).

ehashman replied:
Seems to be missing the right models, I've pinged the release manager on the ticket.

olga-mir commented:
waiting on this PR: https://github.com/Azure/azure-sdk-for-python/pull/10332

arrownj commented on src/azure-cli/requirements.py3.Linux.txt:
```diff
@@ -31,7 +31,7 @@ azure-mgmt-compute==10.0.0
 azure-mgmt-containerinstance==1.5.0
 azure-mgmt-containerregistry==3.0.0rc8
-azure-mgmt-containerservice==8.1.0
+azure-mgmt-containerservice==8.2.0
```
Darwin and Windows missed ?

olga-mir replied:
@arrownj this was needed before it was merged to the python sdk. Now this change is gone.

arrownj commented on src/azure-cli/azure/cli/command_modules/acs/commands.py:
```diff
@@ -119,6 +119,7 @@ def load_command_table(self, _):
     with self.command_group('openshift', openshift_managed_clusters_sdk,
                             client_factory=cf_openshift_managed_clusters) as g:
         g.custom_command('create', 'openshift_create', supports_no_wait=True)
+        g.custom_command('update', 'openshift_update', supports_no_wait=True)
```
Do we have a `wait` command as there are so many `support_to_wait` commands ?

olga-mir replied:
@arrownj yes we do have `wait` command: https://github.com/Azure/azure-cli/blob/67d91648c5cf92cd560db06e1bc62cdcf7480004/src/azure-cli/azure/cli/command_modules/acs/commands.py#L126

olga-mir commented:
@ehashman @kwoodson sdk has now been updated, I will resume work on this PR soon. 
https://github.com/Azure/azure-sdk-for-python/pull/10464/files

olga-mir commented:
@julienstroheker I have tested the `update` command, but in our subscription I am not able to test `az openshift create`

olga-mir commented:
`create` is not working properly. 

I am running 
`az openshift create -n olga-cli-3 -g olga-cli-3 -l eastus2 --private-cluster --management-subnet-cidr 10.0.1.0/24 --debug`

The payload sent is:
`
{"location": "eastus2", "properties": {"openShiftVersion": "v3.11", "networkProfile": {"vnetCidr": "10.0.0.0/8", "managementSubnetCidr": "10.0.1.0/24"}, "routerProfiles": [{"name": "default"}], "masterPoolProfile": {"count": 3, "vmSize": "Standard_D4s_v3", "subnetCidr": "10.0.0.0/24", "apiProperties": {"privateApiServer": true}}, "agentPoolProfiles": [{"name": "compute", "count": 4, "vmSize": "Standard_D4s_v3", "subnetCidr": "10.0.0.0/24", "osType": "Linux", "role": "compute"}, {"name": "infra", "count": 3, "vmSize": "Standard_D4s_v3", "subnetCidr": "10.0.0.0/24", "osType": "Linux", "role": "infra"}], "authProfile": {"identityProviders": [{"name": "Azure AD", "provider": {"kind": "AADIdentityProvider", "clientId": "<id>", "secret": "<secret>", "tenantId": "<id>"}}]}}} 
`
The payload looks correct, but the resulting cluster is not a private cluster.

@julienstroheker any thoughts?

olga-mir commented:
I ran one more check and I want to provide some more detail. 

I took the above payload and placed in the script that creates production cluster using rest API call directly. 
I ran this PR `az openshif create` as described above. 

Using curl the created cluster does appear to be private, while the one created with cli is not (like I observed yesterday).

In the resource group of the cluster that is created with cli, there is  PLS, PLS NIC and `lb-apiserver-internal` are not created. 

When describing both clusters using `az openshift show` I observed following differences:
```
"fqdn": "10.0.0.254" vs "osaxxxxxxxxxxxxxxxxxxxx.eastus2.cloudapp.azure.com",
"publicHostname": "10.0.0.254" vs "publicHostname": "openshift.xxxxxxxxxxxxxxxxxxxx.eastus2.azmosa.io",
```

Because the resulting API that this PR code sends is valid, I assume that the problem is with Azure RP.

olga-mir commented:
one more thing that I've noticed is that the CLI cluster has 3 deployments on the RG, while cluster created with curl had 4. Missing deployment is `pid-<uuid>`

julienstroheker commented:
Your payload is sent with the wrong API Version 2019-09-30-preview, should be 2019-10-27-preview

This needs to be modified inside the _client_factory.py

You can run any command with the `--debug` flag in it to have more details such as payload, headers etc... this is very useful when developing.

Thanks

olga-mir commented:
thanks @julienstroheker indeed I've missed the api version string update (https://github.com/Azure/azure-cli/pull/12240/commits/bc513c8188222c57be1a2dca6cd7b193da6d8f86) I thought models upgrade did the trick. 

I have fixed that and tested again the created cluster is a valid cluster and update refresh-cluster also works. I have rebased and tests passed, so marking this ready for review.

olga-mir commented:
@fengzhou-msft @arrownj can either of you please take another look at this PR, it now looks good from our side.

arrownj commented:
Hi @fengzhou-msft , could you help take a look at it once more? If no problem, I will merge it.
